### PR TITLE
[inductor] fix max_pool2d with ceil mode

### LINKED ISSUE #94775: `torch.compile` returns inconsistent value `MaxPool` when `ceil_mode = True`
### 🐛 Describe the bug

`torch.compile` returns inconsistent value `MaxPool` when `ceil_mode = True`

```py
import torch

torch.manual_seed(420)
x = torch.randn(1, 1, 6, 7)

def func(x):
    y = torch.nn.functional.max_pool2d(x, 1, stride=(2, 2), padding=0, ceil_mode=True) 
    return y

print(func(x))
# tensor([[[[-0.0070,  0.6704,  0.0302, -0.5131],
#           [ 2.0182, -0.2523, -0.0646,  0.0938],
#           [ 0.8388, -0.5796,  1.4694, -1.8402]]]])

print(torch.compile(func)(x))
# tensor([[[[-0.0070,  0.6704,  0.0302, -0.5131],
#           [ 2.0182, -0.2523, -0.0646,  0.0938],
#           [ 0.8388, -0.5796,  1.4694, -1.8402],
#           [   -inf,    -inf,    -inf,    -inf]]]])
```

When `ceil_mode = False`, the results are the same

### Versions

```
Collecting environment information...
PyTorch version: 2.0.0.dev20230213
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: version 3.24.1
Libc version: glibc-2.35

Python version: 3.9.15 (main, Nov 24 2022, 14:31:59)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 515.86.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] functorch==0.2.1
[pip3] numpy==1.23.4
[pip3] torch==2.0.0.dev20230213
[pip3] torchaudio==2.0.0.dev20230213
[pip3] torchtriton==2.0.0+0d7e753227
[pip3] torchvision==0.15.0.dev20230213
[conda] _tflow_select             2.3.0                       mkl  
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] functorch                 0.2.1                    pypi_0    pypi
[conda] libblas                   3.9.0            12_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            12_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            12_linux64_mkl    conda-forge
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] numpy-base                1.23.4           py39h31eccc5_0  
[conda] pytorch                   2.0.0.dev20230213 py3.9_cuda11.7_cudnn8.5.0_0    pytorch-nightly
[conda] pytorch-cuda              11.7                 h778d358_3    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torch                     1.14.0.dev20221207+cu117          pypi_0    pypi
[conda] torchaudio                0.13.0.dev20220923+cu116          pypi_0    pypi
[conda] torchtriton               2.0.0+0d7e753227          pypi_0    pypi
[conda] torchvision               0.14.0.dev20220923+cu116          pypi_0    pypi
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by Valentine233**
Fixes #94775

When ceil mode turns on, max_pool2d has a bug allowing a sliding window to be entirely off bounds. This PR restricts sliding windows to start within the input or left padding.

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Review by desertfire — APPROVED** [review]

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -2697,7 +2697,9 @@ def pooling_size(x, i, kernel_size, stride, padding, ceil_mode):
         x_alt = ir.FloorDiv(
             x + 2 * padding[i] - (kernel_size[i] - 1) + 2 * (stride[i] - 1), stride[i]
         )
-
+        if V.graph.sizevars.size_hint((x_alt - 1) * stride[i] - x - padding[i]) >= 0:
```
Comment (eellison) [inline torch/_inductor/lowering.py:2700]: I dont think this is quite right. If the check here affects correctness we should be adding a guard for it

Comment (jgong5) [reply torch/_inductor/lowering.py:2700]: So, after we do `x_alt -= 1`, do `V.graph.sizevars.guard_leq(0, x_alt * stride[i] - x - padding[i])`?

Comment (Valentine233) [reply torch/_inductor/lowering.py:2700]: Modified. I think maybe we could do `V.graph.sizevars.guard_lt((x_alt - 1) * stride[i] - x - padding[i], 0)` here.

Comment (Valentine233) [reply torch/_inductor/lowering.py:2701]: Modified.

**Review by jgong5 — APPROVED** [review]

**Comment by Valentine233** [comment]
@eellison Could you pls kindly review this PR again? Thanks :)
